larpandoracontent/LArControlFlow/TestBeamCosmicRayTaggingTool.cc

```diff
@@ -0,0 +1,580 @@
+
+void TestBeamCosmicRayTaggingTool::CheckIfTopToBottom(const CRCandidateList &candidates, PfoToBoolMap &pfoToIsCosmicRayMap) const
+{
+    for (const CRCandidate &candidate : candidates)
+    {
+        // Move on if this candidate has already been tagged
+        if (pfoToIsCosmicRayMap.at(candidate.m_pPfo))
+            continue;
+        const float upperY(
+            (candidate.m_endPoint1.GetY() > candidate.m_endPoint2.GetY()) ? candidate.m_endPoint1.GetY() : candidate.m_endPoint2.GetY());
+        const float lowerY(
+            (candidate.m_endPoint1.GetY() < candidate.m_endPoint2.GetY()) ? candidate.m_endPoint1.GetY() : candidate.m_endPoint2.GetY());
+
+        pfoToIsCosmicRayMap[candidate.m_pPfo] = ((upperY > m_face_Yt - m_marginY) && (lowerY < m_face_Yb + m_marginY));
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TestBeamCosmicRayTaggingTool::CheckIfTopEntering(const CRCandidateList &candidates, PfoToBoolMap &pfoToIsCosmicRayMap) const
+{
+    for (const CRCandidate &candidate : candidates)
+    {
+        // Move on if this candidate has already been tagged
+        if (pfoToIsCosmicRayMap.at(candidate.m_pPfo))
+            continue;
+        const float upperY(
+            (candidate.m_endPoint1.GetY() > candidate.m_endPoint2.GetY()) ? candidate.m_endPoint1.GetY() : candidate.m_endPoint2.GetY());
+        pfoToIsCosmicRayMap[candidate.m_pPfo] = (upperY > m_face_Yt - m_marginY);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TestBeamCosmicRayTaggingTool::CheckIfInVetoedTPC(const CRCandidateList &candidates, PfoToBoolMap &pfoToIsCosmicRayMap) const
+{
+    if (m_vetoedTPCs.empty())
+        return;
+
+    for (const CRCandidate &candidate : candidates)
+    {
+        // Move on if this candidate has already been tagged
+        if (pfoToIsCosmicRayMap.at(candidate.m_pPfo))
+            continue;
+
+        // We have to go via the hit volume id to get the TPC
+        CaloHitList caloHitList3D;
+        LArPfoHelper::GetCaloHits(candidate.m_pPfo, TPC_3D, caloHitList3D);
+
+        float maxYPosition{-1.f * std::numeric_limits<float>::max()};
+        unsigned int maxYTPCId{std::numeric_limits<unsigned int>::max()};
+        for (const CaloHit *const pCaloHit : caloHitList3D)
+        {
+            const float hitYPos(pCaloHit->GetPositionVector().GetY());
+            if (maxYPosition < hitYPos)
+            {
+                maxYPosition = hitYPos;
+                const unsigned int maxYDriftVol =
+                    reinterpret_cast<LArCaloHit *>(const_cast<void *>(pCaloHit->GetParentAddress()))->GetLArTPCVolumeId();
+                const unsigned int maxYAPA = reinterpret_cast<LArCaloHit *>(const_cast<void *>(pCaloHit->GetParentAddress()))->GetDaughterVolumeId();
```

Comment:
Can you just opt for a single cast:

`const LArCaloHit *const pLArCaloHit{dynamic_cast<const LArCaloHit *>(pCaloHit->GetParentAddress())};`

and then get the tpc and daughter volumes from that? Appreciate this might still need a reinterpret rather than dynamic, but presumably the const cast can be avoided if assigning to a const LArCaloHit, and the cast in general can be done just once?

Reply:
This single cast didn't work, but I can at least avoid two awful lines